javascript/0041-first-missing-positive.js

```diff
@@ -1,47 +1,46 @@
- /**
-   * Time O(n) | Space O(1)
+/**
+ * https://leetcode.com/problems/first-missing-positive/
+ * Time O(n) | Space O(1)
  * @param {number[]} nums
  * @return {number}
  */
 var firstMissingPositive = function(nums) {
     //-1 1 3 4
-    const swap = (a,b) => {
+    const swap = (a, b) => {
         const temp = nums[a];
         nums[a] = nums[b];
         nums[b] = temp;
     }
+ 
     // swap positions
-    for(let i = 0; i < nums.length; i++) {
+    for (let i = 0; i < nums.length; i++) {
         const el = nums[i];
         const chair = el - 1;
-        if(el >= 1 &&  el <= nums.length + 1 && nums[chair] !== el) {
-             swap(chair, i);
-                i--; // this decrement is important // check this input [3,4,-1,1]
+        if (el >= 1 && el <= nums.length + 1 && nums[chair] !== el) {
+            swap(chair, i);
+            i--; // this decrement is important // check this input [3,4,-1,1]
         }
     }
 
-
-    for(let i = 0; i < nums.length; i++) {
-        if(nums[i] !== i+1) return i+1;
+    for (let i = 0; i < nums.length; i++) {
+        if (nums[i] !== i + 1) return i + 1;
     } 
 
     return nums.length + 1;
 };
 
 /**
- * https://leetcode.com/problems/first-missing-positive/
```

Comment:
The question gives a hint based on the constraints. Need to sort with Time(N) by using cyclic sort.

```js
/**
 * Cyclic Sort
 * Time O(N) | Space O(1)
 * @param {number[]} nums
 * @return {number}
 */
var firstMissingPositive = (nums) =>{
  cyclicSort(nums);

  return search(nums);
};

const cyclicSort = (nums, index = 0) => {
  while (index < nums.length) {
    const num = nums[index];
    const indexKey = (num - 1);
    const indexNum = nums[indexKey];

    if (canSwap(nums, num, indexNum)) {
      swap(nums, index, indexKey);
      continue;
    }

    index += 1;
  }
}

const search = (nums) => {
  for (let index = 0; (index < nums.length); index++) {
    const num = nums[index];
    const indexKey = (index + 1);

    if (!isEqual(num, indexKey)) return indexKey;
  }

  return (nums.length + 1);
}

const canSwap = (nums, num, indexNum) =>
  isPositive(num) &&
  isInBound(num, nums) &&
  !isEqual(num, indexNum);

const swap = (nums, index, indexKey) => 
  [nums[index], nums[indexKey]] = [nums[indexKey], nums[index]];

const isPositive = (num) => (0 < num);

const isInBound = (num, nums) => (num <= nums.length);

const isEqual = (num, indexNum) => (num === indexNum);

```

Reply:
@aadil42 ^